cpp/src/integrations/cudf/partition.cpp

```diff
@@ -78,37 +83,45 @@ std::unordered_map<shuffler::PartID, PackedData> partition_and_pack(
         auto splits = std::vector<cudf::size_type>(
             static_cast<std::uint64_t>(num_partitions - 1), 0
         );
-        return split_and_pack(table, splits, stream, mr);
+        return split_and_pack(table, splits, stream, br);
     }
     auto [reordered, split_points] = cudf::hash_partition(
-        table, columns_to_hash, num_partitions, hash_function, seed, stream, mr
+        table,
+        columns_to_hash,
+        num_partitions,
+        hash_function,
+        seed,
+        stream,
+        br->device_mr()
     );
     std::vector<cudf::size_type> splits(split_points.begin() + 1, split_points.end());
-    return split_and_pack(reordered->view(), splits, stream, mr);
+    return split_and_pack(reordered->view(), splits, stream, br);
 }
 
 std::unordered_map<shuffler::PartID, PackedData> split_and_pack(
     cudf::table_view const& table,
     std::vector<cudf::size_type> const& splits,
     rmm::cuda_stream_view stream,
-    rmm::device_async_resource_ref mr,
+    BufferResource* br,
     std::shared_ptr<Statistics> statistics
 ) {
     RAPIDSMPF_NVTX_FUNC_RANGE();
     RAPIDSMPF_MEMORY_PROFILE(statistics);
     std::unordered_map<shuffler::PartID, PackedData> ret;
-    auto packed = cudf::contiguous_split(table, splits, stream, mr);
+    auto packed = cudf::contiguous_split(table, splits, stream, br->device_mr());
     for (shuffler::PartID i = 0; static_cast<std::size_t>(i) < packed.size(); i++) {
         auto pack = std::move(packed[i].data);
-        ret.emplace(i, PackedData(std::move(pack.metadata), std::move(pack.gpu_data)));
+        ret.emplace(
+            i, PackedData(std::move(pack.metadata), std::move(pack.gpu_data), stream, br)
```

Comment:
Previously we would never spill at this point, now there is an opportunity to do so?

Reply:
Ah ok, it's only `br->move`s with a memory reservation that might spill?

Reply:
yes!